doc/user-guide/testing.rst

```diff
@@ -0,0 +1,272 @@
+By feeding this strategy for dictionaries into the `dims` argument of xarray's `dataarrays` strategy, we can generate
+arbitrary ``DataArray`` objects whose dimensions will always match these specifications.
+
+
+Creating Duck-type Arrays
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Xarray objects don't have to wrap numpy arrays, in fact they can wrap any array type which presents the same API as a
+numpy array (so-called "duck array wrapping", see :ref:`internals.duck_arrays`).
+
+Imagine we want to write a strategy which generates arbitrary `DataArray` objects, each of which wraps a
+``sparse.COO`` array instead of a ``numpy.ndarray``. How could we do that? There are two ways:
+
+1. Create a xarray object with numpy data and use ``.map()`` to convert the underlying array to a
+different type:
+
+.. ipython:: python
+    :okexcept:
+
+    import sparse
+    import hypothesis.extra.numpy as npst
+
+.. ipython:: python
+    :okexcept:
+
+    def convert_to_sparse(da):
+        if da.ndim == 0:
+            return da
+        else:
+            da.data = sparse.COO.from_numpy(da.values)
+            return da
+
+.. ipython:: python
+    :okexcept:
+
+    sparse_dataarrays = xrst.dataarrays().map(convert_to_sparse)
+
+    sparse_dataarrays.example()
+    sparse_dataarrays.example()
+
+2. Pass a strategy which generates the duck-typed arrays directly to the ``data`` argument of the xarray
+strategies:
+
+.. ipython:: python
+    :okexcept:
+
+    @st.composite
+    def sparse_arrays(draw) -> st.SearchStrategy[sparse._coo.core.COO]:
+        """Strategy which generates random sparse.COO arrays"""
+        shape = draw(npst.array_shapes())
+        density = draw(st.integers(min_value=0, max_value=1))
+        return sparse.random(shape, density=density)
+
+.. ipython:: python
+    :okexcept:
+
+    sparse_dataarrays = xrst.dataarrays(data=sparse_arrays())
+
+    sparse_dataarrays.example()
+    sparse_dataarrays.example()
+
+Either approach is fine, but one may be more convenient than the other depending on the type of the duck array which you
+want to wrap.
+
+Creating datasets can be a little more involved. Using method (1) is simple:
+
+.. ipython:: python
+    :okexcept:
+
+    def convert_ds_to_sparse(ds):
+        return ds.map(convert_to_sparse)
+
+.. ipython:: python
+    :okexcept:
+
+    sparse_datasets = xrst.datasets().map(convert_ds_to_sparse)
+
+    sparse_datasets.example()
+
+but building a dataset from scratch (i.e. method (2)) requires building the dataset object in such as way that all of
+the data variables have compatible dimensions. You can build up a dictionary of the form ``{var_name: data_variable}``
+yourself, or you can use the ``data_vars`` argument to the ``data_variables`` strategy (TODO):
+
+.. ipython:: python
+    :okexcept:
+
+    sparse_data_vars = xrst.data_variables(data=sparse_arrays())
+    sparse_datasets = xrst.datasets(data_vars=sparse_data_vars)
+
+    sparse_datasets.example()
```

Comment:
@keewis do you have any thoughts on this section? Given that half the point of this PR is to facilitate testing the duck array wrapping.

I'm worried that currently whilst it's easy to generate a `DataArray` with duck-typed data, it's actually kind of hard to generate a `Dataset` with duck-typed data. I'm trying to think what the minimum set of extra kwargs I would need to add to `data_variables`/`datasets` to make this easier would be. The issue is that you can't just pass `data=sparse_arrays()` to `datasets`, because `sparse_arrays()` will then generate un-alignable variables. For this to work you would have to pass something like a callable which accepts a shape and produces a sparse array of that size, so that I can use `dim_sizes` to create only alignable duck arrays...

Reply:
Passing a callable is what I tried with the `create` class method in #4972, but I'm not sure that is actually the best solution in this case (creating the final strategy becomes a bit complicated). Instead, I wonder if we would actually need some kind of strategy object that can be specialized after having been instantiated. For example, with this:
```python
@st.composite
def pint_arrays(draw, *, shape=None, dtype=None, units=None):
    if shape is None:
        shape = shapes()

    if dtype is None:
        dtype = dtypes()

    if units is None:
        units = units()

    arrays = npst.arrays(shape, dtype)

    return pint.Quantity(draw(arrays), draw(units))
```
we would be able to "pin" the `shape`:
```python
strategy = pint_arrays()

dim_sizes = ...

specialized_strategy = strategy.pin(shape=dim_sizes)
```
In other words, instead of calling `.filter` later we'd modify the input parameters of the composite strategy (looking at the code of `CompositeStrategy` this actually does not seem too complicated to implement).

@Zac-HD, what do you think? Does that make sense to you, or would you recommend to solve this problem differently?

Reply:
I'd be strongly in favor of accepting a callable taking `*, shape, dtype, ...` kwargs and returning a `st.SearchStrategy[DuckArray]]`.  We've got remarkably good support for the Array API standard which will make this increasingly easy to handle ([pass `array_strategy_fn=make_strategies_namespace(duckarray_module).arrays`](https://hypothesis.readthedocs.io/en/latest/numpy.html#array-api)).  Or, of course, `array_strategy_fn=pint_arrays` as per the example above - note the lack of parens!

Then internally, you pass in the shape, dtype, and maybe elements arguments, and then `draw()` from the resulting strategy. 
 This *is* a little awkward, but it's an awkward situation and I've found that focussing on principled composability is *really* valuable when designing strategies.

The `.pin()` solution is much harder than it sounds - users might try to supply strategies created with `st.builds()` or `.map()` or something else, and then you've broken the uniformity of the API: if a strategy is required, *any* strategy that produces reasonable values ought to be acceptable.  Also, bluntly, I expect that the Hypothesis internals would break this for you at some point and then you're in a nasty bind.

Reply:
I had intended to push `.pin` in some form upstream, but I of course forgot about the other types of strategies so I can see why that would not be desirable.

Putting the code into the definition of the composite strategy is much better than what I had before (constructing the examples using `data.draw` directly in the test), so that would be fine with me.

Do you know if it is possible to use `make_strategies_namespace` with additional parameters to the array's constructor, like units for `pint` or chunks for `dask`? I guess if we use the `pint_arrays` function from above we could use `partial` for this (and anyway, `pint` does not implement `__array_namespace__` at the moment).

Reply:
> Do you know if it is possible to use `make_strategies_namespace` with additional parameters to the array's constructor, like units for `pint` or chunks for `dask`? I guess if we use the `pint_arrays` function from above we could use `partial` for this (and anyway, `pint` does not implement `__array_namespace__` at the moment).

`make_strategies_namespace()` does not currently support additional parameters, largely because [the array creation functions](https://data-apis.org/array-api/latest/API_specification/creation_functions.html) don't support them.  I think there would be some forward-compatibility issues with accepting `**kwargs` too, though perhaps not prohibitive.  I can see why that would be desirable though, and I'd be very happy to support it if standardized or just written into a draft.

For Pint, I'd write something like the following:
```python
def pint_arrays(draw, *, shape, dtype, units=units(), array_strategy_fn=npst.arrays):
    return st.builds(pint.Quantity, array_strategy_fn(shape=shape, dtype=dtype), units=units)
```
and then, as you say, use `array_strategy_fn=partial(pint_arrays, units=...)` to customize.

Where possible, I prefer `return st.builds(...)` to `@st.composite` because the latter has to re-run any conditionals or argument validation every time it generates an example.  This is usually just an aesthetic preference though.